rand_distr/src/multinomial.rs

```diff
@@ -0,0 +1,252 @@
+        f.write_str(match self {
+            Error::ProbabilityNegative => "One of the weights is negative or Nan",
+            Error::SumOverflow => "Sum of weights overflows to inf",
+            Error::SumZero => "Sum of weights is zero",
+        })
+    }
+}
+
+/// The [Multinomial](https://en.wikipedia.org/wiki/Multinomial_distribution) distribution `Multinomial(n, w)`.
+#[derive(Debug)]
+pub struct Multinomial {}
+
+impl Multinomial {
+    /// Constructs a new `Multinomial` distribution which samples `K` samples.
+    ///
+    /// `n` is the number of draws.
+    ///
+    /// `weights` have to be non negative and will be normalized to 1.
+    ///
+    /// `K` has to be known at compile time
+    pub fn new_const<const K: usize, I>(
+        n: I,
+        weights: &[f64; K],
+    ) -> Result<MultinomialConst<K, I>, Error>
+    where
+        I: num_traits::PrimInt,
+        u64: num_traits::AsPrimitive<I>,
+        I: num_traits::AsPrimitive<u64>,
+    {
+        let all_pos = weights.iter().all(|&x| x >= 0.0);
+
+        if !all_pos {
+            return Err(Error::ProbabilityNegative);
+        }
+
+        let sum: f64 = weights.iter().sum();
+
+        if !sum.is_finite() {
+            return Err(Error::SumOverflow);
+        }
+
+        if sum == 0.0 {
+            return Err(Error::SumZero);
+        }
+
+        Ok(MultinomialConst::<K, I> { n, weights, sum })
+    }
+
+    #[cfg(feature = "alloc")]
+    /// Constructs a new `Multinomial` distribution which samples `K` samples.
+    ///
+    /// `n` is the number of draws.
+    ///
+    /// `weights` have to be not negative and will be normalized to 1.
+    ///
+    /// `K` can be specified at runtime
+    pub fn new_dyn<I>(n: I, weights: &[f64]) -> Result<MultinomialDyn<'_, I>, Error> {
+        let all_pos = weights.iter().all(|&x| x >= 0.0);
+
+        if !all_pos {
+            return Err(Error::ProbabilityNegative);
+        }
+
+        let sum: f64 = weights.iter().sum();
+
+        if !sum.is_finite() {
+            return Err(Error::SumOverflow);
+        }
+
+        if sum == 0.0 {
+            return Err(Error::SumZero);
+        }
+
+        Ok(MultinomialDyn::<I> { n, weights, sum })
+    }
+}
+/// Multinomial Distribution with compile time known number of categories.
+/// Can be created with [Multinomial::new_const].
+#[derive(Debug, Clone, PartialEq)]
+pub struct MultinomialConst<'a, const K: usize, I> {
+    /// number of draws
+    n: I,
+    /// weights for the multinomial distribution
+    weights: &'a [f64; K],
+    /// sum of the weights
+    sum: f64,
+}
+
+#[cfg(feature = "alloc")]
+/// Multinomial Distribution with number of categories known at runtime.
+/// Can be created with [Multinomial::new_dyn].
+#[derive(Debug, Clone, PartialEq)]
+pub struct MultinomialDyn<'a, I> {
+    /// number of draws
+    n: I,
+    /// weights for the multinomial distribution
+    weights: &'a [f64],
+    /// sum of the weights
+    sum: f64,
+}
```

Comment:
I suggest we just use `weights: Vec<f64>` or `Box<[f64]>`, though we could use [`tinyvec`](https://crates.io/crates/tinyvec).